### DISCUSSION TITLE
Deployment & Migration: 02 - Maintenance Mode

### DISCUSSION THREAD

**Discussion by alexbrowng** [directus/directus#7395] [Feature Requests]
_This is a part of a series of proposals related to create a migration system, in which I would like to know your opinion and suggestions._

_My intention is, if it is interesting for the community and the internal team gives the go-ahead, I would start developing it myself._

_Sorry if I have duplicated another idea._

_Thanks in advance._

### Needs

A system that allows to restrict the level of access to the system temporarily.

### Use cases

- When an intervention is made, do not allow any user to access or modify data.
- When a backup is made, allow to read data but not to create, modify or delete data.
- Production environment, allow to read, create, modify and delete data but not to modify the data model, roles or permissions.

### Background

When doing maintenance tasks such as, updating the data model, roles or permissions, cleaning old data or backups there is no easy method from directus to restrict the access level of the application, then, it disables some requests from Nginx by operation type GET, POST, PATCH, DELETE and endpoint.

### Proposal

Create a system that allows to register execution modes to adapt to the needs of each project, in which, a name is indicated and can be restricted at the level of user roles some operations such as:

- Access to the application (login, reset_password ...)
- Read items (GET /items)
- Create items (POST /items)
- Update items (PUT/PATCH /items)
- Delete items (DELETE /items)
- Modify files (POST/PUT/PATCH/DELETE /files)
- Modify data model (POST/PUT/PATCH/DETELE /collections | /fields, /relations)
- Modify permissions (POST/PUT/PATCH/DETELE /roles | /permissions)
- Launch webhook (directus_webhook)
- Send emails
- Login to the App (/admin)

### Minimal version

Add a new page in the App that allows, register the different modes and activate one of them, plus an indicator to show in which mode it is running and a cache system like the one in the data schema for quick access.

**Data model**

directus_modes (id, key, title, description, code (HTTP code), message (HTTP message), icon (optional), color (optional), permissions ({"role_name": ["permission_key_01", "permission_key_02"]...}), active)

**Middleware**

Create an express middleware that allows depending on the user role and active mode to restrict requests by method (GET,POST,PATCH,PUT,DELETE) and patch (/items,/collections, /fields...) if the request does not have permissions return the defined error code and message.

**Mode service**

GET /modes

GET /modes/:modeId

POST /modes

PATCH /modes/:modeId (when one mode is enabled disable the rest)

DELETE /modes/:modeId

### Futuribles

- Create notification for users
- Schedule activation/deactivation of a mode

**Comment by rijkvanzanten** [comment]
I'm wondering if these modes needs to be this advanced to meet the initial use case 🤔 I like the flexibility, but it sounds like quite a bit of configuration overhead to go into a temporary maintenance mode.

Could it be enough to have a single system maintenance mode flag that will effectively disable all mutations? That in turn would also stop webhooks and emails (as those are only on mutations), and file uploads.

**Reply by alexbrowng** [reply]
I agree, it is possible that a more basic initial version could be made, as you say, maybe a checklist with some options to enable/disable some requests for all users regardless of role.

**Reply by rijkvanzanten** [reply]
Yeah! I think we can find a good middle ground where there is a single "Maintenance Mode" that _Just Works™_ which can then be tweaked with (some) options. 

PS, I don't think "Futuribles" is a word, but it fits super well and I might keep using it (I read it as "Future Deliverables = Futurables") 😄

**Reply by alexbrowng** [reply]
Yes, that's a good starting point. How do we continue?

**Reply by alexbrowng** [reply]
I thought the word "Futuribles" existed in English, I checked the Spanish dictionary and it means "That could exist or occur in the future, especially if a certain condition were given".

**Reply by rijkvanzanten** [reply]
So regarding the enabling/disabling of this mode, I'm assuming it'll have to be a flag in the database right? If it were an environment variable, you'd have to restart/redeploy the current version in order to enable maintenance mode, which feels silly 🤔

**Reply by alexbrowng** [reply]
I also thought about an environment variable as a first approach, but I have some project that has more than one instance of directus raised with load balancing and I think it makes very complex, it is better to have a flag in database, for example, three fields in directus_settings, like:

- maintenance_start_at
- maintenance_end_at (optional)
- maintenance_options (JSON/CSV) with the maintenance mode restrictions.

And I thought about caching it as the data schema for quick access, but I understand that not everyone uses directus with a cache system like redis or memcache, plus you have to make sure on every request if it is still active.

**Reply by rijkvanzanten** [reply]
Also the timed start/end could be done in external means, so `maintenance_enabled` could be a boolean, and then you could easily enable/disable it with a PATCH to `/settings` 🤔

**Reply by alexbrowng** [reply]
Much better

**Reply by alexbrowng** [reply]
Is it okay if I start with this? I can do a PR with an initial proposal and we'll work from there, or do you have any other ideas on where or how you want to start?

**Reply by rijkvanzanten** [reply]
Yeah, works for me! Lemme know on Discord when you're diving into the actual settings for `maintenance_options`. I'd love to discuss that a bit more before we dive in too deep

**Reply by alexbrowng** [reply]
that's done, how can I contact you by discord, it doesn't let me write you directly, my user alejandro-tss#7022

**Reply by rijkvanzanten** [reply]
I've send you a friend request 👍🏻

**Reply by nickrum** [reply]
How about using a `maintenance_role` setting instead of `maintenance_options` with a role that will be applied to each user in maintenance mode? That way this feature would gain the power and flexibility of the role setup and we wouldn't have to reinvent the wheel around maintenance options.

**Reply by rijkvanzanten** [reply]
Ohh interesting approach!

**Reply by alexbrowng** [reply]
I'm in

**Reply by rijkvanzanten** [reply]
<details><summary>Deal.</summary><img src="https://media.giphy.com/media/jZAOQbch4ibfQ97VBS/giphy.gif" alt="" /></details>

**Comment by koksikus** [comment]
Very good concept with temporarily unavailable service. Maybe it should be the option to close the login and the option to close the transmission of data "sending them" out of Directus and other solutions previously mentioned. Disabling the possibility of editing from outside of directus (outside the application) - this will make the data editable only in directus and have control that only one user can modify data in one record at a time.  Full data control for someone who needs data in SQL. Something like in baserow.io/. More possibilities to use directus ;)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
